py/client/pydeephaven/session.py

```diff
@@ -352,24 +338,36 @@ def _connect(self):
 
             self.is_connected = True
 
-            self._timeout = int(session_duration.string_value)
+            self._timeout_seconds = int(session_duration.string_value)/1000.0
             if self._never_timeout:
                 self._keep_alive()
 
     def _keep_alive(self):
-        if self.is_connected:
-            if self._keep_alive_timer:
-                self._refresh_token()
-            self._keep_alive_timer = threading.Timer(self._timeout / 2 / 1000, self._keep_alive)
-            self._keep_alive_timer.daemon = True
-            self._keep_alive_timer.start()
-
-    def _refresh_token(self):
+        trace(f'_keep_alive')
+        if not self.is_connected:
+            return
+        factor = 1
+        ok = True
+        if self._keep_alive_timer:
+            ok = self._refresh_token()
+            if not ok:
+                factor = 0.5
+        timer_wakeup = self._timeout_seconds * factor
+        trace(f'_keep_alive timer_wakeup={timer_wakeup}')
+        self._keep_alive_timer = threading.Timer(timer_wakeup, self._keep_alive)
+        self._keep_alive_timer.daemon = True
+        self._keep_alive_timer.start()
+        if not ok:
+            logging.warning(f'{self._logpfx} Will retry auth token refresh in {timer_wakeup} seconds.')
+
+    def _refresh_token(self) -> bool:
+        trace('_refresh_token')
         try:
-            self._flight_client.authenticate(self._auth_handler)
-        except Exception as e:
-            self.is_connected = False
```

Comment:
It seems that we remove a condition where we deem a session as not usable anymore. It might be OK but does seem to go against FEFF.

Reply:
That's not accurate, this will result in a retry, the session is no unusable at this point.

I did see cases in my testing (which may have been the result of the bug) where one refresh auth attempt failed, and the next one succeeded.  In general any attempt could fail on a network glitch, like any individual RPCs can fail, and should be retried.

Reply:
I was arguing that the _keep_alive would keep on trying forever and never mark the session as 'no longer connected'. Maybe it is OK but usually we would impose a limit on the number of retries.

Reply:
Ok, fair, I added a limit to retries that would revert to the old behavior after a max.